src/pagination.rs

```diff
@@ -0,0 +1,312 @@
+            self.previous_index = index - 1;
+        }
+
+        if index < total_indexes {
+            // we have a next index
+            self.next_index = index + 1;
+            self.next_index_permalink = Some(interpret_permalink(&config, &doc, index + 1)?);
+        }
+        Ok(())
+    }
+}
+
+impl Into<liquid::value::Object> for Paginator {
+    fn into(self) -> liquid::value::Object {
+        let mut object = liquid::value::Object::new();
+        // if no pages, means we have indexes instead, `tags` like cases for exemple
+        if let Some(pages) = self.pages {
+            object.insert("pages".into(), liquid::value::Value::Array(pages));
+        }
+        // list of available indexes, in `tags` like cases
+        if let Some(indexes) = self.indexes {
+            object.insert(
+                "indexes".into(),
+                liquid::value::Value::Array(
+                    indexes.iter().map(liquid::value::Value::scalar).collect(),
+                ),
+            );
+        }
+        object.insert(
+            "index".into(),
+            liquid::value::Value::scalar(self.index as i32),
+        );
+        object.insert(
+            "index_permalink".into(),
+            liquid::value::Value::scalar(self.index_permalink),
+        );
+        if let Some(previous_index_permalink) = self.previous_index_permalink {
+            object.insert(
+                "previous_index".into(),
+                liquid::value::Value::scalar(self.previous_index as i32),
+            );
+            object.insert(
+                "previous_index_permalink".into(),
+                liquid::value::Value::scalar(previous_index_permalink),
+            );
+        }
+        if let Some(next_index_permalink) = self.next_index_permalink {
+            object.insert(
+                "next_index".into(),
+                liquid::value::Value::scalar(self.next_index as i32),
+            );
+            object.insert(
+                "next_index_permalink".into(),
+                liquid::value::Value::scalar(next_index_permalink),
+            );
+        }
+        object.insert(
+            "first_index_permalink".into(),
+            liquid::value::Value::scalar(self.first_index_permalink),
+        );
+        object.insert(
+            "last_index_permalink".into(),
+            liquid::value::Value::scalar(self.last_index_permalink),
+        );
+        object.insert(
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc.front.pagination.as_ref().unwrap();
```

Comment:
Please replace the `unwrap` with an `expect`

`unwrap`s are ok if the condition that guarentees and explains the unwrap is close to the unwrap.  As the distance grows, the need for an `expect` grows.